Harvard-ATG/hxat
# Canvas RCE integration

This PR  allows instructors to embed/insert an annotation assignment object in a Canvas page using the rich content editor. Here's a [brief demo](https://drive.google.com/file/d/1Zfqnyyc804nRY1di2mxIy9ZiQuFJmiyZ/view?usp=sharing) of how it works.

Changes:
- Added XML config to enable a button in the WYSIWYG editor ([examples](https://canvas.instructure.com/doc/api/file.tools_xml.html)), that when clicked, triggers an LTI  `ContentItemSelectionRequest`.
- Updated `middleware.py` to validate a `ContentItemSelectionRequest` message which is similar to an LTI launch message. This triggers the [LTI Content Item Selection Request Workflow](https://www.imsglobal.org/specs/lticiv1p0-intro) ([spec](https://www.imsglobal.org/specs/lticiv1p0/specification)).
- Created views `embed_lti` and `embed_lti_response` that allow the user to select an annotation object and insert it as an LTI link in the content. 
- Created forms `EmbedLtiSelectionForm` and `EmbedLtiResponseForm` to support the selection of content items and the POST back to the LMS with the selected items.
- Updated LTI launch view to accept `collection_id` and `object_id` as custom GET parameters in addition to POST parameters, which will route directly to the annotation view. 

Limitations:
- The course and user must already exist in the DB. If they don't, the user will get a message saying they need to login to the tool first (e.g. via the course navigation). We can/should improve this, but it will require refactoring some logic that is tightly woven into the `launch_lti` view.
- When the tool is embedded on a page,`resource_link_id == context_id` where `context_id` is the course-level identifier. This means we can't use the _starting resource configuration_ method of routing learners to a particular assignment object like we do with module items, but instead must pass the assignment and target IDs directly (`collection_id` and `object_id`).
- The `resource_link_id` is used by the `notifications` django app to identify websocket channels, which will be problematic if there are multiple LTI tools on the same page. For right now, we will advise content authors to only embed on tool per page.

To test:
1. Install in the course with the updated XML config that enables the editor button functionality.
2. Add at least one annotation assignment with a text, video, or image object. To do that, you'll need to open the tool from the left-hand course navigation, and then click *Create Assignment* and then *Add Source* (to add an existing object).
3. Create a Canvas page.
4. When editing the Canvas page, click on the "plug" icon on the toolbar, which should then show "Add from AnntoationsX" as an option. Click that, which should then launch the embed workflow.

@dodget @Tanvez My plan is to do an initial review here (in our fork), then close this PR after making any suggested changes and then re-submit upstream for the HX folks to review and hopefully merge. Since this is quite large, I'd like to just focus on the "core" functionality, so any comments in those areas would be much appreciated:

- File: `hx_lti_initializer/views.py` --> `embed_lti` and `embed_lti_response` view functions
- File: `tests/hx_lti_initializer/test_lti_embed.py`
- File: `tests/hx_lti_initializer/test_lti_forms.py`
